---
title: "MOFA on longitudinal allo-SCT microbiome data"
subtitle: "MOFA model which includes metabolites, plots for manuscript"
author: "Paul Heinrich"
output: 
  html_document:
    toc: TRUE
---

# Notes

Metabolite data was included in the model. No scaling of omics views to unit variance. 
Fitting was performed with medium convergence and 10 factors. No subsets of view features were
selected, so views have a substantial inequality in the number of features (especially
metabolites, with only around 30 features).

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}

library("here")
library("ggplot2")
library("ggpubr")
library("ggrepel")
library("corrplot")
library("rstatix")
library("stringr")
library("patchwork")
library("MOFA2")
library("knitr")
library("tidyr")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("grid")
library("gridExtra")
library("readxl")

here::i_am("analysis/hep62703_2022-10-19_meta_F-10_C-medium_no_scaling/reports/mofa_plots_manuscript.Rmd")

source(here("functions/load_mofa_tmp_copy.R"))

save_data = FALSE

today <- Sys.Date()

set.seed(42)

```

```{r load_data, echo=FALSE, message = FALSE}

analysis_path <- here("analysis/hep62703_2022-10-19_meta_F-10_C-medium_no_scaling")

dir_model <- file.path(analysis_path, "data", "models")
name_model <- "model_metabolites_medium_F10.hdf5"

plot_path <- file.path(analysis_path, "img")
table_path <- file.path(analysis_path, "tables")

model <- load_mofa_tmp_copy(
    file_directory = dir_model,
    file_name = name_model,
    name_tmp_folder = "p107_hp_mofa"
)

os_trm_relapse_updated <- read.csv(
    here("metadata/OS_TRM_Relapse_2022-11-15.csv")
)
colnames(os_trm_relapse_updated) <- c(
    "Project_ID",
    "pat_survival_overall_2yr",
    "pat_TRM_2yr",
    "pat_relapse_2yr"
)

metabolite_index <- read_excel(here("metadata/metabolite_index_5.xlsx"))
contig_table <- read_excel(here("metadata/BCOAT-1.xlsx"))
bcoat_expression_table <- read_excel(here("metadata/BCOAT_qPCR.xlsx"))

name_mapping <- list()
mapping_views <- c("16S", "ITS")

metabolite_name_mapper <- function(metabolite_ids) {
    metabolite_names <- sapply(
        metabolite_ids,
        function(x) paste0(
            toupper(substr(x, 1, 1)),
            substr(x, 2, nchar(x))
        )
    )
    metabolite_names[
        metabolite_names == "Indole_3_carboxyaldehyde"
    ] <- "ICA"
    metabolite_names[
        metabolite_names == "2_methylbutyric_acid"
    ] <- "2-Methylbutyric acid"
    metabolite_names[
        metabolite_names == "3_dehydrocholic_acid"
    ] <- "3-Dehydrocholic acid"
    metabolite_names[
        metabolite_names == "7_dehydrocholic_acid"
    ] <- "7-Dehydrocholic acid"
    metabolite_names[
        metabolite_names == "7_ketolithocholic_acid"
    ] <- "7-Ketolithocholic acid"
    metabolite_names[
        metabolite_names == "G_muricholic_acid_hyocholic_acid"
    ] <- "g-Muricholic acid hyocholic acid"

    metabolite_names <- stringr::str_replace_all(
        metabolite_names, "_", replacement = " "
    )
    names(metabolite_names) <- metabolite_ids

    return(metabolite_names)
}

for (view in mapping_views) {
    filename <- paste0("taxonomic_mapping_", view, ".csv")
    name_mapping[[view]] <- read.csv(
        here("metadata/taxonomic_classification", filename),
        header=TRUE,
        row.names = 1
    )
    mapping_vec <- name_mapping[[view]][["Genus"]]
    names(mapping_vec) <- rownames(name_mapping[[view]])
    name_mapping[[view]] <- mapping_vec
}
name_mapping[["metabolites"]] <- metabolite_name_mapper(
    rownames(get_weights(model)[["metabolites"]])
)

```

```{r get_model_data, echo=FALSE}

views <- views_names(model)
metadata <- samples_metadata(model)
factor_values <- get_factors(model, factors = "all")$group1
factor_weights <- get_weights(model)
model_input <- get_data(model)

# Remove outlier samples (time points too long after allo-SCT)

outlier_samples <- c("R-AS-8", "R-AN-7")
metadata <- metadata[!(metadata[["sample"]] %in% outlier_samples), ]

# Remove double d14 sample of M-AG

metadata <- metadata[!(metadata[["sample"]] %in% "M-AG-3"), ]

outlier_samples <- c("R-AS-8", "R-AN-7")
metadata <- metadata[!(metadata[["sample"]] %in% outlier_samples), ]

# Fix wrong GvHD entry

metadata[metadata[["sample"]] == "M-AV-1", "pat_GvHD"] <- 2

# Add updated overall survival

metadata <- merge(
    x = metadata,
    y = os_trm_relapse_updated,
    by.x = "pat_Project_ID",
    by.y = "Project_ID",
    all.x = TRUE
)

# Add metabolite index

metadata <- merge(
    x = metadata,
    y = metabolite_index,
    by.x = "pat_Project_ID",
    by.y = "ID",
    all.x = TRUE
)

# Add contig table

# Check difference/overlap of samples between metadata and contig table

print("Samples unique to contig table")
setdiff(contig_table[["Project ID"]], metadata[["Project_ID"]])
print("Samples unique to metadata")
setdiff(metadata[["Project_ID"]], contig_table[["Project ID"]])
print("Intersect")
intersect(metadata[["Project_ID"]], contig_table[["Project ID"]])

# Sample R-BB-2 occurs twice in contig table with different vir seq ids. Presumably, the occurence of R-BB-2 at d -7
# (BAX1_V) is actually R-BB-1

kable(metadata[
    metadata[["pat_Project_ID"]] == "R-BB",
    c("sample", "amp_Timepoint", "vir_Sample_ID", "vir_seq_sample_id")
])

contig_table[
    contig_table[["contig_id_vs2"]] == "BAX1_V",
    "Project ID"
] <- "R-BB-1"

# Join contig table

contig_table_reduced <- contig_table[, c("Project ID", "VC-1", "VC-2")]
colnames(contig_table_reduced) <- c("Project_ID", "BCoAT_VC_1", "BCoAT_VC_2")

metadata <- merge(
    x = metadata,
    y = contig_table_reduced,
    by.x = "sample",
    by.y = "Project_ID",
    all.x = TRUE
)

# Add BCoAT qPCR data

metadata$sample[!(metadata$sample %in% bcoat_expression_table$ProjectID)]
bcoat_expression_table$ProjectID[!(bcoat_expression_table$ProjectID %in% metadata$sample)]

bcoat_expression_table_reduced <- bcoat_expression_table %>% select(ProjectID, `copiesBCoAT per gram dry feces`)
colnames(bcoat_expression_table_reduced) <- c("Project_ID", "copies_BCoAT_per_g_dry_feces")

metadata <- merge(
    x = metadata,
    y = bcoat_expression_table_reduced,
    by.x = "sample",
    by.y = "Project_ID",
    all.x = TRUE
)

# Create custom column for metadata analysis

metadata[["pat_GvHD_binary"]] <- as.numeric(metadata[["pat_GvHD"]] > 0)
metadata[["has_BCoAT_VC_1"]] <- as.numeric(metadata[["BCoAT_VC_1"]] > 0)
metadata[["has_BCoAT_VC_2"]] <- as.numeric(metadata[["BCoAT_VC_2"]] > 0)
metadata[["has_BCoAT_expression"]] <- as.numeric(metadata[["copies_BCoAT_per_g_dry_feces"]] > 0)

metadata_factors <- merge(
    x = metadata,
    y = factor_values,
    by.x = "sample",
    by.y = "row.names"
)

timepoints_to_exclude <- c(
    "allo",
    "idx"
)

metadata_factors_subset <- metadata_factors[
    !(metadata_factors[["pre_post_allo_SCT"]] %in% timepoints_to_exclude),
]

metadata_factors_subset[["pre_post_allo_SCT"]] <- as.factor(
    metadata_factors_subset[["pre_post_allo_SCT"]]
)

factor_names <- colnames(metadata_factors_subset)[
    grepl("Factor", colnames(metadata_factors_subset))
]

factors <- as.numeric(str_extract(factor_names, "[0-9]+$"))
names(factors) <- factor_names

clinical_factors <- c(
    "pat_GvHD_binary", "pat_TRM_2yr", "pat_survival_overall_2yr", "atb_Antibiotic_therapy", "Index5binary",
    "has_BCoAT_VC_1", "has_BCoAT_VC_2", "has_BCoAT_expression"
)
clinical_factors_mapping <- c(
    "GvHD", "2-year-TRM", "2-year-Survival", "ABX", "Metabolite index",
    "BCoAT VC-1", "BCoAT VC-2", "BCoAT expression"
)
names(clinical_factors_mapping) <- clinical_factors
clinical_factors_labels <- list(
    c("No GI-GvHD", "GI-GvHD"),
    c("No 2-yr-TRM", "2-yr-TRM"),
    c("2-yr Survival", "No 2-yr survival"),
    c("No ABX", "ABX"),
    c("low", "high"),
    c("no", "yes"),
    c("no", "yes"),
    c("no", "yes")
)
names(clinical_factors_labels) <- clinical_factors
clinical_metadata_colors <- c("#0F7FFE", "#FB0106")

pre_post_values <- unique(metadata_factors_subset[["pre_post_allo_SCT"]])

relevantFactors <- factors[c(7)]
relevantViews <- views[c(2, 3)]

```

```{r factor_plot_function, echo=FALSE}

boxplot_custom <- function(
    df, x, y, color = NULL,
    title = NULL,
    alpha = 1
) {
    colorParam <- color
    if(is.null(color)) {
        color <- x
    }

    plt <- ggplot(
        df,
        mapping = aes_string(
            x = x, y = y
        )
    ) + ggtitle(title) +
    geom_boxplot(
        aes_string(color = color),
        outlier.shape = NA,
        alpha = alpha
    ) +
    theme(
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black")
    )

    if(!is.null(colorParam)) {
        plt <- plt + geom_point(
            aes_string(color = color),
            #size = 0.5,
            position = position_jitterdodge()
        )
    } else {
        plt <- plt + geom_point(
            #size = 0.5,
            position = position_jitter()
        )
    }

    return(plt)
}

```

# Variance analysis

```{r variance_analysis, echo=FALSE}

plt_var_expl <- plot_variance_explained(model, x = "view", y = "factor")
plt_var_expl_abs <- plot_variance_explained(
    model,
    x ="view",
    y = "factor",
    plot_total = TRUE
)[[2]]

plt_var_expl
plt_var_expl_abs

```

# Factor weight plots

```{r factor_weight_plots, echo=FALSE, results="asis"}

factor_weight_plots <- list()

for (factor in names(factors)) {
    factor_weight_plots[[factor]] <- list()
    for (view in views) {
        plt <- plot_top_weights(model,
            view = view,
            factor = factors[factor],
            nfeatures = 15,
            scale = FALSE
        )

        if(view %in% names(name_mapping)) {
            feature_ids <- as.character(plt$data[["feature_id"]])
            feature_ids <- str_remove(feature_ids, paste0("_", view, "$"))
            plt$data[["feature_name"]] <- name_mapping[[view]][feature_ids]

            plt <- plt + scale_x_discrete(
                breaks=plt$data[["feature_id"]],
                labels=plt$data[["feature_name"]]
            )
        }

        factor_weight_plots[[factor]][[view]] <- plt
    }
}

for (factor in names(relevantFactors)) {
    cat(paste0("## ", factor))
    cat("\n\n")
    for (view in relevantViews) {
        cat(paste0("### View: ", view))
        cat("\n\n")
        plot(factor_weight_plots[[factor]][[view]])
        cat("\n\n")
    }
}

```

# Feature-feature correlation heatmaps

```{r factor_feature_correlations, echo=FALSE}

n_weights <- 15
top_weights <- list()

for (factor in names(factors)) {
    top_weights[[factor]] <- list()
    for (view in views) {
        curr_weights <- factor_weights[[view]][, factor]
        weight_sorting <- order(abs(curr_weights), decreasing=TRUE)
        curr_top_weights <- curr_weights[weight_sorting][1:n_weights]
        top_weights[[factor]][[view]] <- curr_top_weights
    }
}

```

```{r feature_feature_correlations, echo=FALSE}

viewCombs <- combn(relevantViews, 2)
colnames(viewCombs) <- apply(
    viewCombs, 2,
    function(x) paste0(x[1], "_", x[2])
)

feature_feature_corr <- list()
feature_feature_corr_p <- list()
feature_feature_corr_p_adj <- list()

for (factor in names(relevantFactors)) {
    feature_feature_corr[[factor]] <- list()
    feature_feature_corr_p[[factor]] <- list()
    feature_feature_corr_p_adj[[factor]] <- list()
    for (comb in colnames(viewCombs)) {
        curr_view_a <- viewCombs[1, comb]
        curr_view_b <- viewCombs[2, comb]
        curr_top_weights_a <- names(top_weights[[factor]][[curr_view_a]])
        curr_top_weights_b <- names(top_weights[[factor]][[curr_view_b]])
        
        corr_mat <- matrix(0, nrow = n_weights, ncol = n_weights)
        rownames(corr_mat) <- curr_top_weights_a
        colnames(corr_mat) <- curr_top_weights_b
        corr_p_val_mat <- corr_mat

        for (feature_a in rownames(corr_mat)) {
            for (feature_b in colnames(corr_mat)) {
                corr <- cor.test(
                    x = model_input[[curr_view_a]]$group1[feature_a, ],
                    y = model_input[[curr_view_b]]$group1[feature_b, ]
                )
                corr_mat[feature_a, feature_b] <- corr$estimate
                corr_p_val_mat[feature_a, feature_b] <- corr$p.value
            }
        }

        corr_p_val_df <- as.data.frame(corr_p_val_mat)
        corr_p_val_df[["view_a"]] <- rownames(corr_p_val_df)
        corr_p_val_long <- pivot_longer(
            corr_p_val_df,
            cols = 1:15,
            names_to = "view_b",
            values_to = "p"
        )
        corr_p_val_long[["p_adj"]] <- p.adjust(
            corr_p_val_long[["p"]], method = "BH"
        )
        corr_p_val_adj_df <- as.data.frame(pivot_wider(
            corr_p_val_long,
            id_cols = "view_a",
            names_from = "view_b",
            values_from = "p_adj"
        ))
        rownames(corr_p_val_adj_df) <- corr_p_val_adj_df[["view_a"]]
        corr_p_val_adj_df <- corr_p_val_adj_df[, colnames(corr_p_val_adj_df) != "view_a"]
        corr_p_val_adj_mat <- as.matrix(corr_p_val_adj_df)

        feature_feature_corr[[factor]][[comb]] <- corr_mat
        feature_feature_corr_p[[factor]][[comb]] <- corr_p_val_mat
        feature_feature_corr_p_adj[[factor]][[comb]] <- corr_p_val_adj_mat
    }
}

```

```{r feature_feature_heatmaps, echo=FALSE, results="asis", fig.dim = c(10, 10)}

feature_feature_heatmaps <- list()
feature_feature_heatmap_matrices <- list()

for (factor in names(feature_feature_corr)) {
    cat(paste0("## ", factor))
    cat("\n\n")
    feature_feature_heatmaps[[factor]] <- list()
    feature_feature_heatmap_matrices[[factor]] <- list()
    for (comb in colnames(viewCombs)) {
        cat(paste0("### Views: ", comb))
        cat("\n\n")
        feature_ids <- list()
        feature_names <- list()
        for (view in viewCombs[, comb]) {
            feature_ids[[view]] <- as.character(
                names(top_weights[[factor]][[view]])
            )
            feature_ids[[view]] <- str_remove(
                feature_ids[[view]], paste0("_", view, "$")
            )

            if(view %in% names(name_mapping)) {
                feature_names[[view]] <- name_mapping[[view]][feature_ids[[view]]]
            } else {
                feature_names[[view]] <- feature_ids[[view]]
            }
        }

        mat <- feature_feature_corr[[factor]][[comb]]
        rownames(mat) <- feature_names[[1]]
        colnames(mat) <- feature_names[[2]]

        hm <- pheatmap(
            mat,
            cluster_rows = FALSE,
            cluster_cols = FALSE,
            breaks=seq(-1, 1, length.out=101),
            silent = TRUE
        )
        feature_feature_heatmaps[[factor]][[comb]] <- hm$gtable
        feature_feature_heatmap_matrices[[factor]][[comb]] <- mat

        corrplot(
            mat,
            col = colorRampPalette(rev(brewer.pal(n = 11, name = "RdBu")))(200),
            tl.col = "black"
        )
        cat("\n\n")
    }
}

```

# Factor box plots with statistical tests, FDR corrected

```{r factor_metadata_statistics, echo=FALSE, results="asis"}

timepoints <- c("7", "14", "21")

metadata_factors_tp <- metadata_factors_subset[
    metadata_factors_subset$amp_Timepoint %in% timepoints,
]
metadata_factors_tp[["amp_Timepoint"]] <- factor(
    metadata_factors_tp[["amp_Timepoint"]], levels = timepoints
)
factors_tp_avg <- metadata_factors_tp %>%
    group_by(pat_Project_ID) %>%
    summarise(across(starts_with("Factor"), mean))

metadata_clin_by_pat <- metadata_factors_tp %>% group_by(pat_Project_ID) %>% summarize(
    across(clinical_factors, function(x) ifelse(!all(is.na(x)), as.numeric(sum(x, na.rm = TRUE) > 0), NA))
)

clinical_factors_standard <- c("pat_GvHD_binary", "pat_TRM_2yr", "pat_survival_overall_2yr", "atb_Antibiotic_therapy", "Index5binary")
clinical_factors_bcoat <- c("has_BCoAT_VC_1", "has_BCoAT_VC_2", "has_BCoAT_expression")

metadata_clin_by_pat <- metadata_clin_by_pat %>% mutate(
    across(clinical_factors_standard, function(x) factor(x, levels = c(0, 1)))
)
metadata_clin_by_pat <- metadata_clin_by_pat %>% mutate(
    across(clinical_factors_bcoat, function(x) factor(x, levels = c(1, 0)))
)

metadata_factors_tp_avg <- merge(
    x = factors_tp_avg,
    y = metadata_clin_by_pat,
    by.x = "pat_Project_ID",
    by.y = "pat_Project_ID"
)

stats_metadata_tp <- list()
p_vals <- c()

# Statistical tests

for (factor in names(relevantFactors)) {
    stats_metadata_tp[[factor]] <- list()

    for (clin in clinical_factors) {
        formula_clin <- formula(paste(factor, "~", clin))

        stats_clin <- metadata_factors_tp_avg %>%
        filter(!is.na(.data[[clin]])) %>%
        rstatix::wilcox_test(formula_clin) %>%
        add_significance("p") %>% add_xy_position(x = clin, dodge = 0.8)

        stats_clin[["y.position"]] <- stats_clin[["y.position"]] + 0.2
        stats_metadata_tp[[factor]][[clin]] <- stats_clin

        p_vals_curr <- stats_clin[["p"]]
        names(p_vals_curr) <- paste(factor, clin, sep = "_")
        p_vals <- c(p_vals, p_vals_curr)
    }
}

# Adjust p values

p_vals_adjusted <- p.adjust(p_vals, method = "BH")

for (factor in names(relevantFactors)) {
    for (clin in clinical_factors) {
        curr_name <- paste(factor, clin, sep = "_")
        stats_metadata_tp[[factor]][[clin]][["p.adj"]] <- signif(p_vals_adjusted[curr_name], 4)
        stats_metadata_tp[[factor]][[clin]] <- stats_metadata_tp[[factor]][[clin]] %>% add_significance("p.adj")
    }
}

```

## Clinical metadata

```{r factor_metadata_plots, echo=FALSE, results="asis"}

factor_metadata_plots_tp <- list()

for (factor in names(relevantFactors)) {
    cat(paste0("### ", factor))
    cat("\n\n")
    factor_metadata_plots_tp[[factor]] <- list()

    for (clin in clinical_factors_standard) {
        cat(paste0("#### Clinical variable: ", clinical_factors_mapping[clin]))
        cat("\n\n")

        stats_clin <- stats_metadata_tp[[factor]][[clin]]
        metadata_factors_tp_avg_subset <- metadata_factors_tp_avg %>% filter(!is.na(.data[[clin]]))

        plt <- boxplot_custom(
            df = metadata_factors_tp_avg_subset[, c(factor, clin)],
            x = clin,
            y = factor,
            color = clin
        ) + labs(
            x = clinical_factors_mapping[clin],
            color = clinical_factors_mapping[clin]
        ) + scale_x_discrete(
            breaks = c(0, 1),
            labels = clinical_factors_labels[[clin]]
        ) + scale_color_manual(
            breaks = c(0, 1),
            labels = clinical_factors_labels[[clin]],
            values = clinical_metadata_colors
        ) +
        #stat_pvalue_manual(stats_clin, label = "p = {p} ({p.signif})", tip.length = 0.02)
        stat_pvalue_manual(stats_clin, label = "p = {p.adj} ({p.adj.signif})", tip.length = 0.02)

        factor_metadata_plots_tp[[factor]][[clin]] <- plt
        print(plt)
        cat("\n\n")
    }
}

```

## BCoAT data

```{r factor_metadata_plots_bcoat, echo=FALSE, results="asis"}

for (factor in names(relevantFactors)) {
    cat(paste0("### ", factor))
    cat("\n\n")
    for (clin in clinical_factors_bcoat) {
        cat(paste0("#### Clinical variable: ", clinical_factors_mapping[clin]))
        cat("\n\n")

        stats_clin <- stats_metadata_tp[[factor]][[clin]]
        metadata_factors_tp_avg_subset <- metadata_factors_tp_avg %>% filter(!is.na(.data[[clin]]))

        plt <- boxplot_custom(
            df = metadata_factors_tp_avg_subset[, c(factor, clin)],
            x = clin,
            y = factor,
            color = clin
        ) + labs(
            x = clinical_factors_mapping[clin],
            color = clinical_factors_mapping[clin]
        ) + scale_x_discrete(
            breaks = c(0, 1),
            labels = clinical_factors_labels[[clin]]
        ) + scale_color_manual(
            breaks = c(0, 1),
            labels = clinical_factors_labels[[clin]],
            values = rev(clinical_metadata_colors)
        ) +
        #stat_pvalue_manual(stats_clin, label = "p = {p} ({p.signif})", tip.length = 0.02)
        stat_pvalue_manual(stats_clin, label = "p = {p.adj} ({p.adj.signif})", tip.length = 0.02)

        factor_metadata_plots_tp[[factor]][[clin]] <- plt
        print(plt)
        cat("\n\n")
    }
}

```

```{r save_data, echo=FALSE, eval=save_data}

# Factor weight plots

for (factor in names(relevantFactors)) {
    for (view in views) {
        plot_name <- paste0(factor, "_weights_", view, ".png")
        plt <- plot(factor_weight_plots[[factor]][[view]])

        ggsave(
                filename = file.path(
                    plot_path,
                    "manuscript",
                    "factor_weights",
                    plot_name
                ),
                plot = plt,
                width = 8,
                height = 8,
                dpi = 300
        )
    }
}

# Factor weight plot data (for Andreas to find family names associated with "unclassified")

for (factor in names(relevantFactors)) {
    for (view in views) {
        file_name <- paste0(factor, "_weights_", view, ".csv")
        factor_weight_data <- factor_weight_plots[[factor]][[view]]$data

        write.csv(
            factor_weight_data,
            file = file.path(table_path, "manuscript", "factor_weights", file_name),
            quote = FALSE,
            row.names = FALSE
        )
    }
}

# Feature-feature correlation heatmaps

for (factor in names(feature_feature_corr)) {
    for (comb in colnames(viewCombs)) {
        plot_name <- paste0(factor, "_feature_feature_corr_", comb, ".png")

        png(filename = file.path(
                plot_path,
                "manuscript",
                "feature_feature_corr",
                plot_name
            ), width = 10, height = 10, unit = "in", res = 100
        )
        corrplot(
            feature_feature_heatmap_matrices[[factor]][[comb]],
            col = colorRampPalette(rev(brewer.pal(n = 11, name = "RdBu")))(200),
            tl.col = "black"
        )
        dev.off()
    }
}

# Feature-feature correlation p-value (adjusted) tables

for (factor in names(feature_feature_corr_p_adj)) {
    for (comb in colnames(viewCombs)) {
        file_name <- paste0(factor, "_feature_feature_corr_p_adj_", comb, ".csv")

        feature_ids <- list()
        feature_names <- list()
        for (view in viewCombs[, comb]) {
            feature_ids[[view]] <- as.character(
                names(top_weights[[factor]][[view]])
            )
            feature_ids[[view]] <- str_remove(
                feature_ids[[view]], paste0("_", view, "$")
            )

            if(view %in% names(name_mapping)) {
                feature_names[[view]] <- name_mapping[[view]][feature_ids[[view]]]
            } else {
                feature_names[[view]] <- feature_ids[[view]]
            }
        }

        mat <- feature_feature_corr_p_adj[[factor]][[comb]]
        rownames(mat) <- feature_names[[1]]
        colnames(mat) <- feature_names[[2]]

        write.csv(
            mat,
            file = file.path(table_path, "manuscript", "feature_feature_corr", file_name),
            quote = FALSE,
            row.names = TRUE
        )
    }
}

# Factor boxplots with fdr-controlled statistics

for (factor in names(relevantFactors)) {
    for (clin in clinical_factors) {
        plot_name <- paste0(
            factor, "_metadata_boxplot_post-avg_", clin, "_stats_fdr.png"
        )
        plt <- factor_metadata_plots_tp[[factor]][[clin]] + labs(x = NULL, color = NULL)

        ggsave(
            filename = file.path(
                plot_path,
                "manuscript",
                "factor_metadata_boxplots",
                plot_name
            ),
            plot = plt,
            width = 8,
            height = 8,
            dpi = 300
        )
    }
}

```

```{r,eval=TRUE}
sessionInfo()
```
